Dev/cakelle2 gcc geos 12.4.0

christophkeller opened the pull request:
A few fixes / updates, as suggested by Mike (my comments are added in **bold**:

    Hi Christoph,
    I recently accepted the Pull Request that you initiated, based on 12.4.0
    Your comments included this line:
    This updates the GEOS-Chem code in GEOSCHEMchem_GridCompMod to GEOS-Chem version 12.4.0. That code is identical to branch 'GEOS_12.4.0' in the GEOS-Chem git repository as of Aug 28, 2019 (--> https://github.com/geoschem/geos-chem.git and https://github.com/geoschem/gchp.git).
    
    Now I'm revisiting the code, to make sure that all of Eric's changes made it in.
    Can you please incoporate these changes into the next release?
    
    File:  Standard.eqn
    Please make the following changes  (you can diff with /discover/nobackup/efleming/Icarus-1_0_p2_MEM_GCC-v11-02/ )
    All line numbers refer to the GIT version.
    line 836: 0.240ClO  should be  0.220ClO
         and   0.670OH  should be 0.660OH
    line 849:  2017/02/22  should be 2018/06/21
    line 857:  2014/02/03...  should be 2018/06/21; JPL 15-10;  ELF
    line 1338: 1.000Cl  should be 2.000Cl
      and  2019/06/27...   should be 2014/02/03; Eastham2014; SDE
    line 1340: 2.000Cl  should be 1.000Cl
      and  2014/02/03...   should be 2019/06/27; Eastham2014; SDE, ELF
**done**: this is a zero-diff change as the Standard.eqn file has not been used for the KPP files    

    GEOSCHEMchem_GridComp.rc
       change  LLSTRAT  from  59   to   72,    (or LM ?)
**done**    

    GEOSCHEMchem_GridCompMod.rc    and   GIGC/Chem_GridCompMod.F90
      change LLSTRAT default  to  72   or  LM
**done**
    
    wetscav_mod.F
    Eric had hard-coded  CNVSCL  = 1   per your suggestion;   do we want to add this back in?
**Left it as is. There is a toggle in GEOSCHEMchem_GridComp.rc for the calculation of convective updraft velocity. The default setting is to use the values used in GEOS-Chem (5 m/s over oceans, 10 m/s over land).**
   
    FJX_spec.dat.rc --   N2O lines  159-163  are different.  why?
**Good question! This must have been a merge error with a previous version of FJX_spec.dat.rc - I updated it to the new values**    

    HEMCO_Config.rc  --   OTD-LIS scaling   is not provided for multi resolutions;  and the current numbers are too low (Eric needed
	to increase the values to get 6 TG of N)
    We also need multi-resolutions  for mass tuning factor.   The setup script uncomments the line corresponding to the desired resolution.
**Not sure I understand this. The scale-aware scale factors is something that we need to address**

     GEOSCHEMchem_GridCompMod.F90   -   to turn on Family Transport:
	after
  		USE PHYSCONSTANTS
	add
    		USE Chem_GroupMod                                  ! For Family Transport

	in Initialize_,    around line 2778  (in the GEOS section) add this:

	    	! Turn on Family Transport
    		CALL Init_GCC_Chem_Groups()
**done**

	Also note that you may be able to run Family Transport in the context of the CTM, by calling Pack_Chem_Groups before advection,
	and Unpack_Chem_Groups after advection, like we do in GEOS-CTM.

        -Mike

mathomp4 commented on GEOSCHEMchem_GridComp/GEOSCHEMchem_GridCompMod.F90:
```diff
@@ -1992,7 +1993,7 @@ SUBROUTINE Initialize_( GC, Import, Export, Clock, RC )
     ! Top stratospheric level
     CALL ESMF_ConfigGetAttribute( GeosCF, value_LLSTRAT,         & 
-                                  Default = 59,                     &
+                                  Default = 72,                     &
```
@christophkeller A question. Should this be hardcoded to 72? Or should it be something like LM...or even can it be something like LM (i.e., does it need a `parameter`)? 

If we move to 132 levels, this could be a fun thing to track down.

mathomp4 replied:
Oh. Hmm. It looks like `GEOSCHEMchem_GridComp/GEOSCHEMchem_GridComp.rc` also has a hard-coded 72. RC files we can transmute at run-time, so that's not as crucial, but I'm always wary of hard-coded 72s in GEOS.

mathomp4 replied:
Looks like @mmanyin had the same question above I was just too blind to see it.

christophkeller replied:
Good point. The hardcoded default value is a relict from older versions of GEOS-Chem where chemistry was restricted to 59 levels. I will change the default to LM, I think that makes a lot of sense.

lizziel commented:
GCHP does use HEMCO_Config.rc, but we use offline lightning and dust files by default now and so do not use the OTD-LIS scaling and mass tuning factor. Using the online is an option but right now users are on their own for values if they decide to do this. This is simply because the resolution-dependent scale factors haven't been submitted/approved for inclusion. 

@christophkeller should we include these in GCHP? If you proceed with the scale factor recommendations commented out in the HEMCO_Config.rc, and use the script to automatically set them, perhaps we should do the same in GCHP.

christophkeller commented:
I like the approach with the resolution tokens and will adopt it.
For the LLFASTJX values I'm happy to use more meaningful default values based on the value of LM (i.e., 601 for 72 levels and 1201 for 132 levels). I will add these updates shortly.

lizziel commented:
I am planning on bringing all of Christoph's updates for GEOS into GEOS-Chem 12.7 so they should not need to be added again after update to that version. Regarding the resolution-dependent settings, I will start a conversation with the GCHP Working Group to get those approved for inclusion in GCHP, or at least in the pipeline for inclusion after tweaking.

lizziel commented:
Looking at this closer, as far as I can see from the source code the OTD-LIS scaling is not used in the lightning HEMCO extension. Is there something I am missing?

lizziel commented:
I have confirmed that the OTD-LIS scaling is no longer used in GEOS-Chem 12.4.0. The removal was part of the update from Lee Murray to use offline lightning:

+!  16 Jan 2019 - L. Murray   - Lightning flash densities and convective depths
+!                              for vertical distribution are now calculated offline 
+!                              at the native model resolution and prescribed in 
+!                              HEMCO_Config.rc as the other model meteorology.
+!                              Model flash climatology remains constrained to the 
+!                              LIS/OTD climatology, as described by Murray et al. 
+!                              (2012). Removed obsolete subroutines and code.

Is this removal going to be a problem for GEOS-Chem in GEOS?

mmanyin commented:
I see what you mean, that  OTD-LIS scaling is no longer used in the code.
Murray's comment refers to "Lightning flash densities and convective depths" that are now
prescribed in  HEMCO_Config.rc,  but I did a quick comparison of the old and new versions
of that file and could not identify much that was added.  I'm surprised that HEMCO_Config.rc
still includes lines for  OTD-LIS .  Almost makes me think that we are not looking at the
12.4 version of that file (?).
Christoph, it probably makes sense to delete the OTD-LIS lines, and then I will approve the edit,
and merge.
I leave it to you to decide that the code is able to adapt to the run-time resolution appropriately.

lizziel commented:
The GEOS-Chem CTM uses different emissions files so the RC files in GEOS are manually updated per version. For reference the GCHP full chemistry HEMCO_Config.rc for 12.4.0 is located here: https://github.com/geoschem/gchp/blob/12.4.0/Run/HEMCO_Config.rc_templates/HEMCO_Config.rc.fullchem. I don't seen anything else that should cause problems but we could do some work to make the files more similar, making future diffs easier.

christophkeller commented:
Whoa, thanks for pointing this out. I wasn't aware that the scale factors were removed, so this is a problem for online simulations. This needs to be updated. I will follow up with Lee to see how to proceed.

christophkeller commented:
I added an update that provides a (temporary) solution to the lightning problem. The issue is as follows:
GEOS-Chem now expects as external inputs lightning flash rates (LFRs) and convective cloud top heights, which are then used to compute NOx production and vertical distribution (within the HEMCO lightning module). This is an elegant solution since it makes it easy to calculate the corresponding quantities in another part of the GCM and then pass them on to GEOS-Chem / HEMCO. Megan Damon has been working on a module that computes LFRs using various definitions, and that module will be perfect for this.
Until this module is available, I suggest that we adopt the solution used by the offline CTM, i.e., read the flash rates and convective cloud top heights from an external file (through ExtData). This is the fix I added here. I tested this update and it does now compute lightning NOx emissions that look reasonable.
I also suggest that we keep the scaling factors in HEMCO_Config.rc for now. The reason for this is that I suspect that we will have to use some sort of scaling factors once we use online LFRs (again), unless we figure out how to compute LFRs consistently across all resolutions. So I don't want to take out some code that we might have to re-add very shortly. We can always delete the scale factors later once we know exactly how the new online calculations of LFR work out.
